common/make/internal/executable_building_emscripten.mk

```diff
@@ -0,0 +1,195 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+#
+# This file contains the implementation of the ../executable_building.mk
+# interface that builds using Emscripten (that generates WebAssembly binaries).
+#
+# common.mk must be included before including this file.
+#
+
+
+# Sanity-check that the environment is set up.
+ifeq (,$(EMSDK))
+$(error You must run "source env/activate" in the repository root directory at
+		least once before running "make".)
+endif
+
+# Specifies the build configuration (default "Release", or "Debug").
+CONFIG ?= Release
+
+# A pattern rule for "dir.stamp" files in any directory. The rule allows to
+# abstract away calling "mkdir" in output directories. It's intended to be used
+# as an order-only prerequisite - for example:
+#   some_target: some_prerequisite | some_dir/dir.stamp
+%dir.stamp :
+	mkdir -p $(dir $@)
+	@echo Stamp > $@
+
+# Directory where temporary build artifacts (like .o object files) are stored.
+BUILD_DIR := ./out-artifacts-emscripten/$(CONFIG)
+# Directory where library .a files (built by LIB_RULE) are installed.
+LIB_DIR := $(ROOT_PATH)/env/emsdk/lib/$(CONFIG)
+
+# Macro that returns the .o object file path for the given source file.
+# Note that ".." are replaced with a placeholder "__", so that the object files
+# stay under BUILD_DIR.
+# Arguments:
+# $1 ("SOURCE") - Source file path.
+define OBJ_FILE_NAME
+$(BUILD_DIR)/$(basename $(subst ..,__,$(1))).o
+endef
+
+# Flags passed to the Emscripten compiler/linker tools.
+#
+# Explanation:
+# bind: Enables "Embind" (the technology used by our C++ wrappers for talking
+#   with JavaScript).
+# pthread: Enables Pthreads support (for C/C++ multi-threading, etc.).
+# MODULARIZE: Puts Emscripten module JavaScript loading code into a factory
+#   function, in order to control its loading from other JS code and to avoid
+#   avoid name conflicts with unrelated code.
+# EXPORT_NAME: Name of the JavaScript function that loads the Emscripten module.
+# DYNAMIC_EXECUTION: Disable dynamic code execution in Emscripten JavaScript
+#   code (not doing this will cause "unsafe-eval" Content Security Policy
+#   violations when running this code inside Chrome Apps/Extensions).
+EMSCRIPTEN_FLAGS := \
+  --bind \
+  -pthread \
+  -s MODULARIZE=1 \
+  -s 'EXPORT_NAME="loadEmscriptenModule_$(TARGET)"' \
+  -s DYNAMIC_EXECUTION=0 \
+
+ifeq ($(CONFIG),Release)
+
+# Add flags specific to release builds.
+#
+# Explanation:
+# O3: Enable advanced optimizations.
+EMSCRIPTEN_FLAGS += \
+	-O3 \
+
+else ifeq ($(CONFIG),Debug)
```

Comment:
not related to this line but I find that using the `EMCC_DEBUG` env variable can be helpful while in Debug mode: https://emscripten.org/docs/porting/Debugging.html#debug-mode-emcc-debug